include/onnxruntime/core/framework/sparse_csrcformat_rep.h

```diff
@@ -0,0 +1,140 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
+
+#include "core/framework/sparse_tensor.h"
+
+namespace onnxruntime {
+/// <summary>
+/// This class represents Compressed Storage Row(Column) sparse format
+/// CSS for Row Major Matrices
+/// CSC for Column Major Matrices
+/// </summary>
+class SparseCsrcFormatRep : public SparseRep {
+ public:
+  enum Order {
+    kRowMajor = 1,
+    kColMajor = 2
+  };
+
+  /// <summary>
+  /// Constructor that allocates memory for values and indices
+  /// The indices must be inited separately
+  /// </summary>
+  /// <param name="data_type"></param>
+  /// <param name="values_shape"></param>
+  /// <param name="total_allocation_size"></param>
+  /// <param name="allocator"></param>
+  /// <param name="order"></param>
+  SparseCsrcFormatRep(const MLDataType data_type, const TensorShape& values_shape, 
+                      int64_t total_allocation_size, const AllocatorPtr& allocator, Order order)
+      : SparseRep(data_type, values_shape, total_allocation_size, allocator),
+        major_(order),
+        inner_indices_(),
+        outer_indices_() {}
+
+  /// <summary>
+  /// Constructor that allocates no memory and utilizes user provided buffer
+  /// </summary>
+  /// <param name="data_type"></param>
+  /// <param name="values_shape"></param>
+  /// <param name="values_data"></param>
+  /// <param name="location"></param>
+  /// <param name="order"></param>
+  SparseCsrcFormatRep(MLDataType data_type, const TensorShape& values_shape, void* values_data,
+                      const OrtMemoryInfo& location, Order order)
+      : SparseRep(data_type, values_shape, values_data, location),
+        major_(order),
+        inner_indices_(),
+        outer_indices_() {}
+
+  ~SparseCsrcFormatRep() override;
+
+  ORT_DISALLOW_COPY_AND_ASSIGNMENT(SparseCsrcFormatRep);
+
+  void InitIndices(const TensorShape& inner_shape, void* inner_data,
+                   const TensorShape& outer_shape, void* outer_data);
+
+  void InitIndices(const TensorShape& inner_shape, const TensorShape& outer_shape);
+
```

Comment:
what's the use case where the indicies would need to be altered after the class in constructed? as public accessors to mutate the internals makes it impossible for the class to maintain invariants, I'd prefer not to have them if possible.

naively I would have expected you're either constructing from dense data (and can calculate these values at that time), or you're constructure from sparse data (and are provided these values).

Reply:
I think InitIndices() should have been private, as they are used at construction time only.